logprep/processor/normalizer/processor.py

```diff
@@ -237,105 +238,128 @@ def _apply_rules(self, event, rule):
         be written in a way that no such warnings are produced during normal operation because each
         warning should be an indicator of incorrect rules or unexpected/changed events.
         """
-        self._try_add_grok(event, rule)
-        self._try_add_timestamps(event, rule)
-        for before, after in rule.substitutions.items():
-            self._try_normalize_event_data_field(event, before, after)
+        self._apply_grok(event, rule)
+        self._apply_timestamp_normalization(event, rule)
+        for source_field, target_field in rule.substitutions.items():
+            self._apply_field_copy(event, source_field, target_field)
 
-    def _try_add_grok(self, event: dict, rule: NormalizerRule):
+    def _apply_grok(self, event: dict, rule: NormalizerRule):
+        """
+        Applies the grok pattern of a given NormalizerRule to a given event, by matching the field
+        value against the grok pattern. If no pattern matches a grok failure field is written to the
+        event (if configured).
+        """
+        one_matched = False
+        source_field, source_value = None, None
         for source_field, grok in rule.grok.items():
             source_value = self._get_dotted_field_value(event, source_field)
             if source_value is None:
                 continue
-
-            if self._count_grok_pattern_matches:
-                matches = grok.match(source_value, self._grok_pattern_matches)
-            else:
-                matches = grok.match(source_value)
+            matches = self._get_grok_matches(grok, source_value)
             for normalized_field, field_value in matches.items():
                 if field_value is not None:
                     self._try_add_field(event, normalized_field, field_value)
+                    one_matched = True
+        if not one_matched:
+            self._write_grok_failure_field(event, rule, source_field, source_value)
 
-    def _try_add_timestamps(self, event: dict, rule: NormalizerRule):
-        for source_field, normalization in rule.timestamps.items():
-            timestamp_normalization = normalization["timestamp"]
+    def _get_grok_matches(self, grok, source_value):
+        if self._count_grok_pattern_matches:
+            return grok.match(source_value, self._grok_pattern_matches)
+        return grok.match(source_value)
 
-            source_timestamp = self._get_dotted_field_value(event, source_field)
+    def _write_grok_failure_field(self, event, rule, source_field, source_value):
+        """
+        If grok pattern exist and are configured with a failure target field, then write the source
+        field and the first 100 characters in the event.
+        """
+        if not rule.grok.items():
+            return
+        failure_target_field = rule.grok.get(source_field).failure_target_field
+        if failure_target_field:
+            source_field_path = source_field.replace(".", ">")
```

Comment:
what is going on here? Please add a comment or modify docstring. Alternatively, as this seems to relate to add_field_to, the respective behaviour can be documented there.

Reply:
revised methods docstring